driver/src/orderbook/streamed/state.rs

```diff
@@ -79,6 +113,8 @@ impl State {
     /// `block_batch_id` is the current batch based on the timestamp of the block that contains the
     ///  event.
     pub fn apply_event(mut self, event: &Event, block_batch_id: BatchId) -> Result<Self> {
+        ensure!(self.last_batch_id <= block_batch_id, "event in the past");
```

Comment:
I see a lot of `ensure` happening... is this the same as assert? Are these directly relevant to this PR?

Reply:
It is like an assert but intead of panicking it returns an error.

You can check where the ensures come from. It is at the top `use anyhow::{anyhow, bail, ensure, Result};`. The documentation for that is here https://docs.rs/anyhow/1.0.28/anyhow/macro.ensure.html . Or you can try `cargo doc --open` which generates the documentation locally and then searching for `ensure`.